### PR TITLE
Allow to build with NT6 exports

### PR DESCRIPTION
**Pull request by tkreuzer**
## Purpose

This adds a CMAKE config option to specify the dll export version.
It also fixes a number of issues that prevented it from compiling.

JIRA issue: [CORE-11288](https://jira.reactos.org/browse/CORE-11288)

## Proposed changes

- Add DLL_EXPORT_VERSION config option
- Fix a number of exports in ntdll, kernel32, msvcrt, advapi32
- Add NT6 builders for GitHub

### CODE REVIEW

**Review by learn-more — APPROVED** [review]

**Review by DarkFire01 — APPROVED** [review]

**Review by GeoB99 — APPROVED** [review]

File: dll/win32/kernel32/kernel32.spec [hunk 1]
```diff
@@ -93,15 +93,15 @@
...
 @ stdcall ConvertFiberToThread()
 @ stdcall -stub -version=0x600+ ConvertNLSDayOfWeekToWin32DayOfWeek(long)
 @ stdcall -stub -version=0x600+ ConvertSystemTimeToCalDateTime(ptr long ptr)
 @ stdcall ConvertThreadToFiber(ptr)
 @ stdcall ConvertThreadToFiberEx(ptr long)
 @ stdcall CopyFileA(str str long)
-@ stdcall CopyFileExA(str str ptr ptr ptr long)
-@ stdcall CopyFileExW(wstr wstr ptr ptr ptr long)
+@ stdcall CopyFileExA (str str ptr ptr ptr long)
+@ stdcall CopyFileExW (wstr wstr ptr ptr ptr long)
```
Comment (binarymaster) [inline dll/win32/kernel32/kernel32.spec:104]: There are some spaces added, probably by a mistake (and only in kernel32 commit).

Comment (tkreuzer) [reply dll/win32/kernel32/kernel32.spec]: Fixed

File: dll/win32/msvcrt/msvcrt.spec [hunk 2]
```diff
@@ -241,7 +241,7 @@
 @ cdecl -arch=i386 -norelay __CxxFrameHandler2(ptr ptr ptr ptr) __CxxFrameHandler
 @ cdecl -version=0x600+ -arch=x86_64 -norelay __CxxFrameHandler2(ptr ptr ptr ptr) __CxxFrameHandler
 @ cdecl -arch=arm -norelay __CxxFrameHandler3(ptr ptr ptr ptr)
-@ cdecl -version=0x600+ -arch=i386,x86_64 -norelay __CxxFrameHandler3(ptr ptr ptr ptr)
+#@ cdecl -version=0x600+ -arch=i386,x86_64 -norelay -private __CxxFrameHandler3(ptr ptr ptr ptr)
```
Comment (binarymaster) [inline dll/win32/msvcrt/msvcrt.spec:244]: Commented out intentionally?

**Review by binarymaster — APPROVED** [review]
Few things to quickfix, otherwise LGTM.

- Commit says `Define _NTSYSAPI_` but the diff defines `_NTSYSTEM_`
- You can add [CORE-11288](https://jira.reactos.org/browse/CORE-11288) to commit messages

File: dll/win32/msvcrt/msvcrt.spec (hunk 2 above)
Comment (tkreuzer) [reply dll/win32/msvcrt/msvcrt.spec:244]: yes, that causes linker errors. Will add a comment.

**Review by learn-more — APPROVED** [review]

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -118,7 +133,7 @@ jobs:
       with:
         path: src
     - name: Configure
-      run: cmake -S src -B build -G Ninja -DCMAKE_TOOLCHAIN_FILE:FILEPATH=toolchain-msvc.cmake -DARCH:STRING=${{matrix.arch}} -DCMAKE_BUILD_TYPE=${{matrix.config}} -DENABLE_ROSTESTS=1 -DENABLE_ROSAPPS=1
+      run: cmake -S src -B build -G Ninja -DCMAKE_TOOLCHAIN_FILE:FILEPATH=toolchain-msvc.cmake -DARCH:STRING=${{matrix.arch}} -DCMAKE_BUILD_TYPE=${{matrix.config}} -DENABLE_ROSTESTS=1 -DENABLE_ROSAPPS=1 -DDLL_EXPORT_VERSION=${{matrix.dllver}}
```
Comment (HBelusca) [inline .github/workflows/build.yml:136]: Or "NT_COMPAT_VERSION"? (If it's going to become also the version ReactOS will expose to programs)

File: dll/apisets/api-ms-win-crt-private-l1-1-0.spec [hunk 4]
```diff
@@ -8,23 +8,26 @@
...
-@ stdcall -arch=i386 __CxxFrameHandler2() msvcrt.__CxxFrameHandler2
-@ stdcall -arch=arm __CxxFrameHandler3() msvcrt.__CxxFrameHandler3
+@ cdecl -arch=i386,x86_64 __CxxFrameHandler() msvcrt.__CxxFrameHandler
+@ cdecl -arch=i386 __CxxFrameHandler2() msvcrt.__CxxFrameHandler2
+@ cdecl -version=0x600+ -arch=x86_64 __CxxFrameHandler2() msvcrt.__CxxFrameHandler2
+@ cdecl -arch=arm __CxxFrameHandler3() msvcrt.__CxxFrameHandler3
+#@ cdecl -version=0x600+ -arch=i386,x86_64 __CxxFrameHandler3() msvcrt.__CxxFrameHandler3
 @ stdcall -arch=i386 __CxxLongjmpUnwind() msvcrt.__CxxLongjmpUnwind
-@ stdcall -arch=i386 __CxxQueryExceptionSize() msvcrt.__CxxQueryExceptionSize
+@ stdcall -arch=i386 __CxxQueryExceptionSize() msvcrt.__CxxQueryExceptionSize # should be x64, too
```
Comment (HBelusca) [inline dll/apisets/api-ms-win-crt-private-l1-1-0.spec:30]: Is it also causing a link error if activated also for x64 ?

File: dll/apisets/api-ms-win-crt-private-l1-1-0.spec [hunk 5]
```diff
@@ -40,48 +43,50 @@
...
 @ stub __dcrt_initial_narrow_environment
-@ stub __intrinsic_abnormal_termination
+@ stub -arch=i386 __intrinsic_abnormal_termination
 @ stub __intrinsic_setjmp
 @ stub __intrinsic_setjmpex
 @ stub __processing_throw
 @ stub __report_gsfailure
 @ stub __std_exception_copy
 @ stub __std_exception_destroy
+@ stub -arch=x86_65 __std_terminate
```
Comment (HBelusca) [inline dll/apisets/api-ms-win-crt-private-l1-1-0.spec:53]: 65-bits yay! :partying_face: 
```suggestion
@ stub -arch=x86_64 __std_terminate
```

File: dll/win32/advapi32/advapi32.spec [hunk 6]
```diff
@@ -66,8 +66,10 @@
 @ stdcall CommandLineFromMsiDescriptor(wstr ptr ptr)
 @ stub ComputeAccessTokenFromCodeAuthzLevel
 @ stdcall ControlService(long long ptr)
-@ stdcall ControlTraceA(double str ptr long) ntdll.EtwControlTraceA
-@ stdcall ControlTraceW(double wstr ptr long) ntdll.EtwControlTraceW
+@ stdcall -version=0x502 ControlTraceA(double str ptr long) ntdll.EtwControlTraceA
+@ stdcall -stub -version=0x600+ ControlTraceA(double str ptr long)
+@ stdcall -version=0x502 ControlTraceW(double wstr ptr long) ntdll.EtwControlTraceW
+@ stdcall -stub -version=0x600+ ControlTraceW(double wstr ptr long)
```
Comment (HBelusca) [inline dll/win32/advapi32/advapi32.spec:72]: Do they have a separate implementation done in advapi32 in Vista+? (besides the one in ntdll)
Maybe @DarkFire01 knows more?

**Comment by HBelusca** [comment]
(The "requested change" is just for the x86_65 thing. The rest are just questions. The RtlRemovePrivileges becoming a stub whereas it is already implemented in the code base is weird, though.)

File: .github/workflows/build.yml [hunk 7]
```diff
@@ -8,6 +8,12 @@ jobs:
         compiler: [gcc, clang]
         arch: [i386, amd64]
         config: [Debug, Release]
+        dllver: ['0x502', '0x600']
+        exclude:
+          - dllver: 0x600
+            compiler: clang
```
[the remark this answers is not in the document]
Comment (HBelusca) [reply .github/workflows/build.yml:14]: ok makes sense.

File: dll/apisets/api-ms-win-crt-private-l1-1-0.spec [hunk 8]
```diff
@@ -40,48 +43,50 @@
...
-@ stub _o__CIfmod
-@ stub _o__CIlog
-@ stub _o__CIlog10
-@ stub _o__CIpow
-@ stub _o__CIsin
-@ stub _o__CIsinh
-@ stub _o__CIsqrt
-@ stub _o__CItan
-@ stub _o__CItanh
+@ stub -arch=i386 _o__CIacos
```
[the remark this answers is not in the document]
Comment (tkreuzer) [reply dll/apisets/api-ms-win-crt-private-l1-1-0.spec:74]: no

File: dll/win32/kernel32/winnls/string/lcformat.c [hunk 9]
```diff
@@ -31,6 +31,20 @@
 #include <debug.h>
 DEBUG_CHANNEL(nls);
 
+#undef _WIN32_WINNT
+#undef WINVER
+#define _WIN32_WINNT 0x600
+#define WINVER 0x600
```
[the remark this answers is not in the document]
Comment (tkreuzer) [reply dll/win32/kernel32/winnls/string/lcformat.c:37]: No

File: sdk/cmake/gcc.cmake [hunk 10]
```diff
@@ -415,6 +415,8 @@ function(spec2def _dllname _spec_file)
 
     if(__spec2def_VERSION)
         set(__version_arg "--version=0x${__spec2def_VERSION}")
```
[the remark this answers is not in the document]
Comment (tkreuzer) [reply sdk/cmake/gcc.cmake:417]: If you feel like doing so.

File: .github/workflows/build.yml (hunk 3 above)
Comment (tkreuzer) [reply .github/workflows/build.yml:136]: No, not neccessarily.

**Pull request merged**